crates/runix/src/flake_ref/file.rs

```
@@ -0,0 +1,315 @@
+use std::borrow::Cow;
+use std::fmt::Display;
+use std::str::FromStr;
+
+use derive_more::Deref;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use url::Url;
+
+use self::application::{Application, ApplicationProtocol};
+use super::lock::NarHash;
+use super::protocol::{self, Protocol, WrappedUrl, WrappedUrlParseError};
+use super::{BoolReprs, FlakeRefSource};
+
+pub type FileUrl<Protocol> = WrappedUrl<Protocol>;
+
+/// <https://cs.github.com/NixOS/nix/blob/f225f4307662fe9a57543d0c86c28aa9fddaf0d2/src/libfetchers/tarball.cc#L287>
+#[derive(Deserialize, Serialize, Debug, PartialEq, Eq, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct FileBasedRef<Protocol: FileProtocol, A: ApplicationProtocol> {
+    pub url: WrappedUrl<Protocol>,
+
+    #[serde(rename = "type")]
+    #[serde(bound(deserialize = "Application<A>: Deserialize<'de>"))]
+    #[serde(bound(serialize = "Application<A>: Serialize"))]
+    _type: Application<A>,
+
+    #[serde(flatten)]
+    pub attributes: FileAttributes,
+}
+
+pub type FileRef<Protocol> = FileBasedRef<Protocol, application::File>;
+pub type TarballRef<Protocol> = FileBasedRef<Protocol, application::Tarball>;
+
+#[derive(Deserialize, Serialize, Debug, PartialEq, Eq, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct FileAttributes {
+    #[serde(flatten)]
```

Having both `deny_unknown_fields` and `flatten` is misleading

---

Well technically not, but serde also has some issue with it as i've linked above.
I mean we DO want to deny fields that are not covered by the attributes, like `github:flox/runix?tef=feature/flakeref2` should fail with "unknown attribute 'tef'"...